stacks-signer/src/v0/signer_state.rs

```diff
@@ -981,4 +1017,73 @@ impl LocalStateMachine {
         };
         state.tx_replay_set.clone()
     }
+
+    /// Handle a possible bitcoin fork. If a fork is detetected,
+    /// return the transactions that should be replayed.
+    pub fn handle_possible_bitcoin_fork(
+        &self,
+        db: &SignerDb,
+        client: &StacksClient,
+        expected_burn_block: &NewBurnBlock,
+        prior_state_machine: &SignerStateMachine,
+        is_in_tx_replay_mode: bool,
+    ) -> Result<Option<Vec<StacksTransaction>>, SignerChainstateError> {
+        if expected_burn_block.burn_block_height <= prior_state_machine.burn_block_height
+                && expected_burn_block.consensus_hash != prior_state_machine.burn_block
+                // TODO: handle fork while still in replay
+                && !is_in_tx_replay_mode
+        {
+            info!("Signer State: fork detected";
+                "expected_burn_block.height" => expected_burn_block.burn_block_height,
+                "expected_burn_block.hash" => %expected_burn_block.consensus_hash,
+                "prior_state_machine.burn_block_height" => prior_state_machine.burn_block_height,
+                "prior_state_machine.burn_block" => %prior_state_machine.burn_block,
+            );
+            // Determine the tenures that were forked
+            let mut parent_burn_block_info =
+                db.get_burn_block_by_ch(&prior_state_machine.burn_block)?;
+            let last_forked_tenure = prior_state_machine.burn_block;
+            let mut first_forked_tenure = prior_state_machine.burn_block;
+            let mut forked_tenures = vec![(
+                prior_state_machine.burn_block,
+                prior_state_machine.burn_block_height,
+            )];
+            while parent_burn_block_info.block_height > expected_burn_block.burn_block_height {
+                parent_burn_block_info =
+                    db.get_burn_block_by_hash(&parent_burn_block_info.parent_burn_block_hash)?;
+                first_forked_tenure = parent_burn_block_info.consensus_hash;
+                forked_tenures.push((
+                    parent_burn_block_info.consensus_hash,
+                    parent_burn_block_info.block_height,
+                ));
+            }
+            let fork_info =
+                client.get_tenure_forking_info(&first_forked_tenure, &last_forked_tenure)?;
+            let mut forked_blocks = fork_info
+                .iter()
+                .flat_map(|fork_info| {
+                    fork_info
+                        .nakamoto_blocks
+                        .iter()
+                        .flat_map(|blocks| blocks.iter())
+                })
+                .collect::<Vec<_>>();
```

Comment:
I think this can be simplified with a `flatten`:

```
let mut forked_blocks = fork_info
    .iter()
    .flat_map(|f| f.nakamoto_blocks.iter().flatten())
    .collect::<Vec<_>>();
```